core_impl/src/main/java/com/google/instrumentation/common/EventQueue.java

```diff
@@ -0,0 +1,165 @@
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * A low-latency event queue for background updating of (possibly contended) objects. This is
+ * intended for use by instrumentation methods to ensure that they do not block foreground
+ * activities. To customize the action taken on reading the queue, derive a new class from
+ * {@link EventQueueEntry} and pass it to the {@link #enqueue} method.  The
+ * {@link EventQueueEntry#process} method of your class will be called and executed in a background
+ * thread. This class is a Singleton.
+ *
+ * <p>Example Usage:
+ * Given a class as follows:
+ * <pre>
+ * public class someClass {
+ *   public void doSomething() {
+ *     // Do the work of the method. One result is a measurement of something.
+ *     int measurement = doSomeWork();
+ *     // Make an update to the class state, based on this measurement. This work can take some
+ *     // time, but can be done asynchronously, in the background.
+ *     update(measurement);
+ *   }
+ *
+ *   public void update(int arg) {
+ *     // do something
+ *   }
+ * }
+ * </pre>
+ *
+ * <p>The work of calling {@code someClass.update()} can be executed in the backgound as follows:
+ * <pre>
+ * public class someClass {
+ *   // Add a EventQueueEntry class that will process the update call.
+ *   private static final class SomeClassUpdateEvent implements EventQueueEntry {
+ *     private final SomeClass someClassInstance;
+ *     private final int arg;
+ *
+ *     SomeObjectUpdateEvent(SomeObject someClassInstance, int arg) {
+ *       this.someClassInstance = someClassInstance;
+ *       this.arg = arg;
+ *     }
+ *
+ *     &#064;Override
+ *     public void process() {
+ *       someClassInstance.update(arg);
+ *     }
+ *   }
+ *
+ *   public void doSomething() {
+ *     int measurement = doSomeWork();
+ *     // Instead of calling update() directly, create an event to do the processing, and insert
+ *     // it into the EventQueue. It will be processed in a background thread, and doSomething()
+ *     // can return immediately.
+ *     EventQueue.getInstance.enqueue(new SomeClassUpdateEvent(this, measurement));
+ *   }
+ * }
+ *</pre>
+ */
+
+@ThreadSafe
+public final class EventQueue {
+  // An event in the {@link EventQueue}. Just holds a reference to an EventQueueEntry.
+  private static final class InstrumentationEvent {
+    private EventQueueEntry entry = null;
+
+    // Set the EventQueueEntry associated with this InstrumentationEvent.
+    void setEntry(EventQueueEntry entry) {
+      this.entry = entry;
+    }
+
+    // Get the EventQueueEntry associated with this InstrumentationEvent.
+    EventQueueEntry getEntry() {
+      return entry;
+    }
+  }
+
+  // Factory for InstrumentationEvent.
+  private static final class InstrumentationEventFactory
+      implements EventFactory<InstrumentationEvent> {
+    @Override
+    public InstrumentationEvent newInstance() {
+      return new InstrumentationEvent();
+    }
+  }
+
+  /**
```

Comment:
Don't need javadoc for private class. We decided to do it only for at least package-protected classes/methods.